### ISSUE TITLE
Type narrowing doesn't work inside Record

### ISSUE THREAD

**Issue by AlyCode** [microsoft/TypeScript#51205]
# Bug Report

### 🔎 Search Terms

`explicit type narrowing` `record`

### 🕗 Version & Regression Information

Tested on lastest version `4.8.4` and nighty version `4.9.0-dev.20221017`

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?#code/JYOwLgpgTgZghgYwgAgMIAtgBsAmyDeAsAFDJnIgCuAtgPwBcFNARtANwkC+JJoksiFAAU4UCOAIlyyBJlxiQjAEoQEAeyg4APAGcwUUAHMANGjk4AfB2LdiJGJRAIwwNSGRwsWAOJi4-ABV0OBAAMWAANwgACgAHUXEwRhEFMABKRmY1NSwIEMlSchgNZGj1ED0Zc2RQZHjUgDpZbBwFNILpaWAYUqlO-vrEpvMFAG1m3ABdBqpqZABeReRHHAgYUAg8AB8tvv7pAFZkAFo6hPBhlrGJnGnZ5C155AAGPc72okL98jEwSih3PAsDoINZvrZpBCfhA-gDkPpKKCuCQgA)

### 💻 Code

```ts
interface Child {
    num?: number;
}

interface Parent {
    children: Record<string, Child>;
}

function allGreaterThanFive(parent: Parent): boolean {
    for (const child in parent.children) {
        if (
            parent.children[child].num === undefined ||
            5 - parent.children[child].num <= 0
        ) {
            return false;
        }
    }
    return true;
}

```

### 🙁 Actual behavior

`parent.children[child].num` is being reported that it could be undefined

### 🙂 Expected behavior

Typescript narrowing down that `parent.children[child].num` can't be undefined

**Comment by MartinJohns** [comment]
Duplicate of #10530. Type narrowing does not occur for indexed access forms `e[k]` where `k` is not a literal.

Just store it in a variable (which IMO is much cleaner anyway).
